--- a/net/minecraft/inventory/container/ChestContainer.java
+++ b/net/minecraft/inventory/container/ChestContainer.java
@@ -7,6 +7,7 @@
 import net.minecraft.item.ItemStack;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.inventory.InventoryView;
 
 public class ChestContainer extends Container {
    private final IInventory field_75155_e;
@@ -74,6 +75,13 @@
 
    }
 
+   // CraftBukkit start
+   @Override
+   public InventoryView getBukkitView() {
+      return null;
+   }
+   // CraftBukkit end
+
    public boolean func_75145_c(PlayerEntity p_75145_1_) {
       return this.field_75155_e.func_70300_a(p_75145_1_);
    }
