--- a/net/minecraft/inventory/container/HopperContainer.java
+++ b/net/minecraft/inventory/container/HopperContainer.java
@@ -5,6 +5,7 @@
 import net.minecraft.inventory.IInventory;
 import net.minecraft.inventory.Inventory;
 import net.minecraft.item.ItemStack;
+import org.bukkit.inventory.InventoryView;
 
 public class HopperContainer extends Container {
    private final IInventory field_94538_a;
@@ -36,6 +37,13 @@
 
    }
 
+   // CraftBukkit start
+   @Override
+   public InventoryView getBukkitView() {
+      return null;
+   }
+   // CraftBukkit end
+
    public boolean func_75145_c(PlayerEntity p_75145_1_) {
       return this.field_94538_a.func_70300_a(p_75145_1_);
    }
