--- a/net/minecraft/inventory/container/WorkbenchContainer.java
+++ b/net/minecraft/inventory/container/WorkbenchContainer.java
@@ -18,6 +18,7 @@
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.inventory.InventoryView;
 
 public class WorkbenchContainer extends RecipeBookContainer<CraftingInventory> {
    private final CraftingInventory field_75162_e = new CraftingInventory(this, 3, 3);
@@ -96,6 +97,13 @@
       });
    }
 
+   // CraftBukkit start
+   @Override
+   public InventoryView getBukkitView() {
+      return null;
+   }
+   // CraftBukkit end
+
    public boolean func_75145_c(PlayerEntity p_75145_1_) {
       return func_216963_a(this.field_217070_e, p_75145_1_, Blocks.field_150462_ai);
    }
