--- a/net/minecraft/server/management/PlayerProfileCache.java
+++ b/net/minecraft/server/management/PlayerProfileCache.java
@@ -36,17 +36,19 @@
 import java.util.Locale;
 import java.util.Map;
 import java.util.UUID;
+import java.util.concurrent.LinkedBlockingDeque;
 import javax.annotation.Nullable;
 import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.util.JSONUtils;
 import org.apache.commons.io.IOUtils;
+import org.spigotmc.SpigotConfig;
 
 public class PlayerProfileCache {
    public static final SimpleDateFormat field_152659_a = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss Z");
    private static boolean field_187322_c;
    private final Map<String, PlayerProfileCache.ProfileEntry> field_152661_c = Maps.newHashMap();
    private final Map<UUID, PlayerProfileCache.ProfileEntry> field_152662_d = Maps.newHashMap();
-   private final Deque<GameProfile> field_152663_e = Lists.newLinkedList();
+   private final Deque<GameProfile> field_152663_e = new LinkedBlockingDeque<GameProfile>(); // CraftBukkit
    private final GameProfileRepository field_187323_g;
    protected final Gson field_152660_b;
    private final File field_152665_g;
@@ -144,7 +146,7 @@
          this.field_152663_e.remove(gameprofile);
          this.field_152663_e.addFirst(gameprofile);
       } else {
-         GameProfile gameprofile1 = func_187319_a(this.field_187323_g, s);
+         GameProfile gameprofile1 = func_187319_a(this.field_187323_g, p_152655_1_); // Spigot - use correct case for offline players
          if (gameprofile1 != null) {
             this.func_152649_a(gameprofile1);
             playerprofilecache$profileentry = this.field_152661_c.get(s);
@@ -199,7 +201,7 @@
    }
 
    public void func_152658_c() {
-      String s = this.field_152660_b.toJson(this.func_152656_a(1000));
+      String s = this.field_152660_b.toJson(this.func_152656_a(SpigotConfig.userCacheCap));
       BufferedWriter bufferedwriter = null;
 
       try {
