--- a/net/minecraft/server/management/PlayerProfileCache.java
+++ b/net/minecraft/server/management/PlayerProfileCache.java
@@ -36,6 +36,7 @@
 import java.util.Locale;
 import java.util.Map;
 import java.util.UUID;
+import java.util.concurrent.LinkedBlockingDeque;
 import javax.annotation.Nullable;
 import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.util.JSONUtils;
@@ -46,7 +47,7 @@
    private static boolean field_187322_c;
    private final Map<String, PlayerProfileCache.ProfileEntry> field_152661_c = Maps.newHashMap();
    private final Map<UUID, PlayerProfileCache.ProfileEntry> field_152662_d = Maps.newHashMap();
-   private final Deque<GameProfile> field_152663_e = Lists.newLinkedList();
+   private final Deque<GameProfile> field_152663_e = new LinkedBlockingDeque<GameProfile>(); // CraftBukkit
    private final GameProfileRepository field_187323_g;
    protected final Gson field_152660_b;
    private final File field_152665_g;
